iroh-net/src/lib.rs

```diff
@@ -1,11 +1,102 @@
+//! [`NodeId`] is always a required parameter to establish a connection.
+//!
+//! When accepting connections the peer's [`NodeId`] is authenticated.  However it is up to
+//! the application to decide if a particular peer is allowed to connect or not.
+//!
+//!
+//! # Relay Servers
+//!
+//! Relay servers exist to ensure all iroh-net nodes are always reachable.  They accept
+//! **encrypted** traffic for iroh-net nodes which are connected to them, forwarding it to
+//! the correct destination based on the [`NodeId`] only.  Since nodes only send encrypted
+//! traffic, the Relay servers can not decode any traffic for other iroh-net nodes and only
+//! forwards it.
+//!
+//! The connections to the Relay server are initiated as normal HTTP 1.1 connections using
+//! TLS.  Once connected the transport is upgraded to a plain TCP connection using a custom
+//! protocol.  All further data is then sent using this custom relaying protocol.  Usually
+//! soon after the connection is established via the Relay it will migrate to a direct
+//! connection.  However if this is not possible the connection will keep flowing over the
+//! relay server as a fallback.
+//!
+//! Additionally to providing reliable connectivity between iroh-net nodes, Relay servers
+//! provide some functions to assist in [hole punching].  They have various services to help
+//! nodes understand their own network situation.  This includes offering a [STUN] server,
+//! but also a few HTTP extra endpoints as well as responding to ICMP echo requests.
+//!
+//!
+//! # Connections and Streams
+//!
+//! Connections are managed using the [`Endpoint`].  To establish a connection to an
+//! iroh-net node you need to know three pieces of information:
+//!
+//! - The [`NodeId`] of the peer to connect to.
+//! - Some addressing information:
+//!   - Usually the [`RelayUrl`] identifying the Relay server.
+//!   - Sometimes, or usually additionally, any direct addresses which might be known.
+//! - The QUIC/TLS Application-Layer Protocol Negotiation, or [ALPN], name to use.
+//!
+//! The ALPN is used by both sides to agree on which application-specific protocol will be
+//! used over the resulting QUIC connection.  These can be protocols like `h3` used for
+//! [`HTTP/3`], but more commonly will be a custom identifier for the application.
+//!
+//! Once connected the API exposes QUIC streams.  These are very cheap to create so can be
+//! created at any time and can be used to create very many short-lived stream as well as
+//! long-lived streams.  There are two stream types to choose from:
+//!
+//! - **Uni-directional** which only allows the peer which initiated the stream to send
+//!   data.
+//!
+//! - **Bi-directional** which allows both peers to send and receive data.  However the
+//!   initiator of this stream has to send data before the peer will be aware of this
+//!   stream.
+//!
+//! Additionally to be extremely light-weight, streams can be interleaved and will not block
```

Comment:
```suggestion
//! Additionally, to be extremely light-weight, streams can be interleaved and will not block
```

Reply:
I'm not sure this comma works really.  Maybe it's better to avoid this construction entirely, e.g.:
```suggestion
//! As well as streams being extremely light-weight, they can be interleaved and will not block
```
Would that be better?